client_plugin/drivers/shared/etcdops.go

```diff
@@ -280,3 +292,104 @@ func addrToEtcdClient(addr string) (*etcdClient.Client, error) {
 
 	return cli, nil
 }
+
+// writeVolMetadata - Update or Create volume metadata in KV store
+func (d *VolumeDriver) writeVolMetadata(entries []kvPair) error {
+
+	var ops []etcdClient.Op
+	var msg string
+	var err error
+
+	// Create a client to talk to etcd
+	etcdAPI := d.createEtcdClient()
+	defer etcdAPI.Close()
+
+	// ops hold multiple operations that will be done to etcd
+	// in a single revision
+	for _, elem := range entries {
+		ops = append(ops, etcdClient.OpPut(elem.key, elem.value))
+	}
+
+	// Lets write the metadata in a single transaction
+	// Use a transaction if more than one entries are to be written
+	if len(entries) > 1 {
+		_, err = etcdAPI.Txn(context.TODO()).Then(ops...).Commit()
+	} else {
+		_, err = etcdAPI.Do(context.TODO(), ops[0])
+	}
+
+	if err != nil {
+		msg = fmt.Sprintf("Failed to write metadata. Reason: %v", err)
+		log.Warningf(msg)
+		return errors.New(msg)
+	}
+	return nil
+}
+
+// readVolMetadata - Read volume metadata in KV store
+func (d *VolumeDriver) readVolMetadata(keys []string) ([]kvPair, error) {
+	var entries []kvPair
+	var msg string
+
+	// Create a client to talk to etcd
+	etcdAPI := d.createEtcdClient()
+	defer etcdAPI.Close()
+
+	// Read does not require transactions
+	// Just read all the keys required
+	for _, k := range keys {
+		elem := kvPair{key: k, value: ""}
+		ctx, cancel := context.WithTimeout(context.Background(),
+			requestTimeout)
+		resp, err := etcdAPI.Get(ctx, k)
+		// cancel is context cancelFunc()
+		// Its purpose is to ensure that the previous etcd op has
+		// come to an end. In this case if Get is still running
+		// in background, it will cancel it
+		cancel()
+		if err != nil {
+			msg = fmt.Sprintf("Failed to read metadata. Reason: %v", err)
+			log.Warningf(msg)
+			return entries, errors.New(msg)
```

Comment:
Once entries is declared, even though its nil, its len() will be 0. Caller of this function is supposed to match the len() of received entries to len() of requested keys.

So this will work fine. Verified the same in a separate small piece of code.